### ISSUE TITLE
Too high backoff delay values cause long overflow

### ISSUE THREAD

**Issue by BartoszSiemienczuk** [failsafe-lib/failsafe#215]
While building a network policy with high delay values I receive a long overflow exception. 

```
java.lang.ArithmeticException: long overflow
	at java.lang.Math.multiplyExact(Math.java:892)
	at java.time.Duration.toNanos(Duration.java:1186)
	at net.jodah.failsafe.RetryPolicy.withBackoff(RetryPolicy.java:394)
```

The code to replciate this may be something like this: 
```
 final RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
                .withBackoff(500, Integer.MAX_VALUE, ChronoUnit.MILLIS, 1.50)
```

It then tries to convert that to nanoseconds to compare the values which is causing the overflow. 
```

  public RetryPolicy<R> withBackoff(long delay, long maxDelay, ChronoUnit chronoUnit, double delayFactor) {
   (...)
     Duration delayDuration = Duration.of(delay, chronoUnit);
    Duration maxDelayDuration = Duration.of(maxDelay, chronoUnit);
    Assert.state(maxDuration == null || delayDuration.toNanos() < maxDuration.toNanos(),
      "delay must be less than the maxDuration");
    Assert.isTrue(delayDuration.toNanos() < maxDelayDuration.toNanos(), "delay must be less than the maxDelay");
   (...)
    return this;
  }
```

**Comment by whiskeysierra** [comment]
Out of curiosity, what do you want to do? Endless retries with exponential backoff until it eventually works? Or do you have a maximum number of retries configured?

**Comment by BartoszSiemienczuk** [comment]
I want to have this configurable by end-user and was just trying some default values to set. And yes, I also have sensible max attempts configured along with this :)

**Comment by whiskeysierra** [comment]
> I want to have this configurable by end-user and was just trying some default values to set.

Makes sense. Maybe Failsafe could adopt Duration universally in its API or use BigInteger under the hood instead of primitives.

**Comment by Tembrel** [comment]
Fixing the problem at that point __might__ still leave a problem here (depending on the fix):

https://github.com/jhalterman/failsafe/blob/2afad67d920f8137a914020fc213d428980f9aff/src/main/java/net/jodah/failsafe/RetryPolicyExecutor.java#L169

That call to `toNanos()` could overflow, throwing an exception. Note that the line has some very tricky code besides that: It relies on widening conversion so that the call is to `Math.min(float, float)`. The narrowing conversion of float to long won't throw a runtime exception, and does reasonable things if the result of `Math.min` can't be represented as a `long`, but this is non-obvious stuff that should be documented. The first few times I read it, I completely missed what it was doing.

**Comment by whiskeysierra** [comment]
That's really only a problem as long as everything is normalized to
primitive numbers internally. If it would use duration and it's compare and
arithmetic methods it should be fine.

**Comment by jhalterman** [comment]
Something like Guava's `LongMath.saturatedAdd` is a good example of how to avoid this problem of overflowing longs when doing math.

**Comment by jhalterman** [comment]
I just released 3.0.1 with this change.
